lib/std/http/Server.zig

```diff
@@ -158,11 +161,33 @@ pub const BufferedConnection = struct {
     }
 
     pub fn writeAll(bconn: *BufferedConnection, buffer: []const u8) WriteError!void {
-        return bconn.conn.writeAll(buffer);
+        if (bconn.write_buf.len - bconn.write_end <= buffer.len) {
+            @memcpy(bconn.write_buf[bconn.write_end..], buffer);
+            bconn.write_end += @intCast(u16, buffer.len);
+        } else {
+            try bconn.conn.writeAll(bconn.write_buf[0..bconn.write_end]);
+            bconn.write_end = 0;
+
+            try bconn.conn.writeAll(buffer);
+        }
     }
 
     pub fn write(bconn: *BufferedConnection, buffer: []const u8) WriteError!usize {
-        return bconn.conn.write(buffer);
+        if (bconn.write_buf.len - bconn.write_end <= buffer.len) {
```

Comment:
This write() function will blow up in the @memcpy if you write more that 0x2000 bytes (8k)

Im doing this for now, and it seems to work well enough

```
    pub fn write(bconn: *BufferedConnection, buffer: []const u8) WriteError!usize {
        const buffer_remaining_len = bconn.write_buf.len - bconn.write_end;
        if (buffer_remaining_len <= buffer.len) {
            // data wont fit in the buffer, so fill whats left of the buffer and write it out
            @memcpy(bconn.write_buf[bconn.write_end..], buffer[0..buffer_remaining_len]);
            bconn.write_end += @intCast(u16, buffer_remaining_len);

            try bconn.conn.writeAll(bconn.write_buf[0..bconn.write_end]);
            bconn.write_end = 0;

            return buffer_remaining_len;
        } else {
            try bconn.conn.writeAll(bconn.write_buf[0..bconn.write_end]);
            bconn.write_end = 0;

            return try bconn.conn.write(buffer);
        }
    }
```

Reply:
... and my "fix" above is probably not 100% correct either.  Smells like a subtle off-by-1 error ?

It looks fine to me when trying to reason about it, but when using this modified code, I am seeing a corrupted JSON reply being received by the client app, every dozen-to-couple_of_dozen requests.    My JSON payloads are typically around half a kb+, so that is suspiciously close to the 8K mark where the corruption occurs.

The corruption is usually a few missing chars, or some random unprintable chars in the JSON payload in 1 spot, then it looks fine from there on.  Approx 8k later (??) it corrupts again in the same way.  

Am attempting to debug that further, but it might take a while yet.

Reply:
Fixed it - quick rewrite of std.http.Server.BufferedConnection, and no more corrupted replies.

```
    pub fn writeAll(bconn: *BufferedConnection, buffer: []const u8) WriteError!void {
        _ = try bconn.write(buffer);
    }

    pub fn write(bconn: *BufferedConnection, buffer: []const u8) WriteError!usize {
        const remaining_buf_len: usize = bconn.write_buf.len - bconn.write_end;
        if (buffer.len < remaining_buf_len) {
            @memcpy(bconn.write_buf[bconn.write_end .. bconn.write_end + buffer.len], buffer);
            bconn.write_end += @intCast(u16, buffer.len);
        } else {
            // data wont fit in the buffer, so flush the buffer then  write all the data
            try bconn.flush();
            try bconn.conn.writeAll(buffer[0..]);
        }
        return buffer.len;
    }

    pub fn flush(bconn: *BufferedConnection) WriteError!void {
        defer bconn.write_end = 0;
        return bconn.conn.writeAll(bconn.write_buf[0..bconn.write_end]);
    }
```
Same thing should apply to std.http.Client lines 319 - 347 .... but it might be better to move common BufferedConnection code into something shared instead

Reply:
Also means that can remove the res.connection.flush() from res.do() ... it will get flushed on the next call to res.finish() instead.  So the Headers + Body can be 1 write sys call, instead of a 1 write for the headers, then 1 write for the body.